core/drivers/crypto/aspeed/hace_ast2600.c

```diff
@@ -0,0 +1,358 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Aspeed Technology Inc.
+ */
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include <io.h>
+#include <drvcrypt_hash.h>
+#include <mm/core_mmu.h>
+#include <mm/core_memprot.h>
+#include <kernel/delay.h>
+#include <tee/cache.h>
+
+#include "hace_ast2600.h"
+
+#define HACE_BASE	0x1e6d0000
+
+/* register offsets and bit fields */
+#define HACE_STS		0x1C
+#define   HACE_STS_HASH_INT		BIT(9)
+#define   HACE_STS_HASH_BUSY		BIT(0)
+#define HACE_HASH_DATA		0x20
+#define HACE_HASH_DIGEST	0x24
+#define HACE_HASH_HMAC_KEY	0x28
+#define HACE_HASH_DATA_LEN	0x2C
+#define HACE_HASH_CMD		0x30
+#define   HACE_HASH_CMD_ACCUM		BIT(8)
+#define   HACE_HASH_CMD_ALG_SHA1	BIT(5)
+#define   HACE_HASH_CMD_ALG_SHA256	(BIT(6) | BIT(4))
+#define   HACE_HASH_CMD_ALG_SHA384	(BIT(10) | BIT(6) | BIT(5))
+#define   HACE_HASH_CMD_ALG_SHA512	(BIT(6) | BIT(5))
+#define   HACE_HASH_CMD_SHA_BE		BIT(3)
+
+/* buffer size based on SHA-512 need */
+#define HASH_BLK_BUFSZ	128
+#define HASH_DGT_BUFSZ	64
+
+register_phys_mem(MEM_AREA_IO_SEC,
+		  HACE_BASE,
+		  SMALL_PAGE_SIZE);
+
+struct ast2600_hace_ctx {
+	const struct crypto_hash_ops *ops;
+	uint32_t cmd;
+	uint32_t algo;
+	uint32_t dgt_size;
+	uint32_t blk_size;
+	uint32_t pad_size;
+	uint64_t total[2];
+	uint8_t buf[HASH_BLK_BUFSZ];
+	uint8_t digest[HASH_DGT_BUFSZ] __aligned((8));
+};
+
+static vaddr_t hace_virt;
+struct mutex hace_mtx = MUTEX_INITIALIZER;
+
+static const uint32_t iv_sha1[8] = {
+	0x01234567, 0x89abcdef, 0xfedcba98, 0x76543210,
+	0xf0e1d2c3, 0, 0, 0
+};
+
+static const uint32_t iv_sha256[8] = {
+	0x67e6096a, 0x85ae67bb, 0x72f36e3c, 0x3af54fa5,
+	0x7f520e51, 0x8c68059b, 0xabd9831f, 0x19cde05b
+};
+
+static const uint32_t iv_sha384[16] = {
+	0x5d9dbbcb, 0xd89e05c1, 0x2a299a62, 0x07d57c36,
+	0x5a015991, 0x17dd7030, 0xd8ec2f15, 0x39590ef7,
+	0x67263367, 0x310bc0ff, 0x874ab48e, 0x11155868,
+	0x0d2e0cdb, 0xa78ff964, 0x1d48b547, 0xa44ffabe
+};
+
+static const uint32_t iv_sha512[16] = {
+	0x67e6096a, 0x08c9bcf3, 0x85ae67bb, 0x3ba7ca84,
+	0x72f36e3c, 0x2bf894fe, 0x3af54fa5, 0xf1361d5f,
+	0x7f520e51, 0xd182e6ad, 0x8c68059b, 0x1f6c3e2b,
+	0xabd9831f, 0x6bbd41fb, 0x19cde05b, 0x79217e13
+};
+
+static TEE_Result ast2600_hace_process(struct crypto_hash_ctx *ctx,
+				       const uint8_t *data, size_t len)
+{
+	TEE_Result rc = TEE_SUCCESS;
+	uint32_t sts;
+	uint64_t tref;
+	struct ast2600_hace_ctx *hace_ctx;
+
+	hace_ctx = (struct ast2600_hace_ctx *)(void *)ctx;
```

Comment:
I think the proper way is to use `container_of()`, see `core/drivers/crypto/caam/hash/caam_hash.c`.